src/util/circular_buffer.c

```diff
@@ -0,0 +1,253 @@
+	uint64_t item_count = atomic_fetch_add(&cb->item_count, 1);
+	if (unlikely(item_count >= cb->item_cap)) {
+		cb->item_count = cb->item_cap;
+	}
+
+	// determine current and next write position
+	uint64_t offset = atomic_fetch_add(&cb->write, cb->item_size);
+
+	// check for buffer overflow
+	if (unlikely(cb->data + offset >= cb->end_marker)) {
+		// write need to circle back
+		// [., ., ., ., ., ., A, B, C]
+		//                           ^  ^
+		//                           W0 W1
+		uint64_t overflow = offset + cb->item_size;
+
+		// adjust offset
+		// [., ., ., ., ., ., A, B, C]
+		//  ^  ^
+		//  W0 W1
+		offset -= cb->item_size * cb->item_cap;
+
+		// update write position
+		// multiple threads "competing" to update write position
+		// only the thread with the largest offset will succeed
+		// for the above example, W1 will succeed
+		//
+		// [., ., ., ., ., ., A, B, C]
+		//        ^
+		//        W
+		atomic_compare_exchange_strong(&cb->write, &overflow, offset + cb->item_size);
+	}
+
+	// return slot pointer
+	return cb->data + offset;
+}
+
+// Read oldest item from buffer.
+// This function is not thread-safe.
+// This function pops the oldest item from the buffer.
+void *CircularBuffer_Read(CircularBuffer cb, void *item) {
+	RedisModule_Assert(cb != NULL);
+
+	// make sure there's data to return
+	if (unlikely(CircularBuffer_Empty(cb))) {
+		return NULL;
+	}
+
+	void *read = cb->read;
+
+	// update buffer item count
+	cb->item_count--;
+
+	// copy item from buffer to output
+	if (item != NULL) {
+		memcpy(item, cb->read, cb->item_size);
+	}
+
+	// advance read position
+	// circle back if read reached the end of the buffer
+	cb->read += cb->item_size;
+	if (unlikely(cb->read >= cb->end_marker)) {
+		cb->read = cb->data;
+	}
+
+	// return original read position
+	return read;
+}
+
+// Sets the read pointer to the beginning of the buffer.
+// assuming the buffer looks like this:
+//
+// [., ., ., A, B, C, ., ., .]
+//                    ^
+//                    W
+//
+// CircularBuffer_ResetReader will set 'read' to A
+//
+// [., ., ., A, B, C, ., ., .]
+//           ^        ^
+//           R        W
+//
+void CircularBuffer_ResetReader(CircularBuffer cb) {
+	// compensate for circularity
+	uint64_t write = cb->write;
+
+	// compute newest item index, e.g. newest item is at index k
+	uint idx = write / cb->item_size;
+
+	// compute offset to oldest item
+	// oldest item is n elements before newest item
+	//
+	// example:
+	//
+	// [C, ., ., ., ., ., ., A, B]
+	//
+	// idx = 1, item_count = 3
+	// offset is 1 - 3 = -2
+	//
+	// [C, ., ., ., ., ., ., A, B]
+	//     ^                 ^
+	//     W                 R
+
+	int offset = idx - cb->item_count;
+	offset *= cb->item_size;
+
+	if (offset >= 0) {
+		// offset is positive, read from beginning of buffer
+		cb->read = cb->data + offset;
+	} else {
+		// offset is negative, read from end of buffer
+		cb->read = cb->end_marker + offset;
+	}
```

Comment:
worth adding an assert that the read offset did not exceed the buffer boundaries.

Reply:
How can this happen?
The `item_count` is aligned with `write` pointer, and this function should be called under a lock, such that no writes are happening concurrently.